### ISSUE TITLE
[Feature request] Support not equal to filter

### ISSUE THREAD

**Issue by SagareGanesh** [typesense/typesense#257]
It will be good if we provide a filter for not equal conditions.

Example

```
  product_schema = {
     "name" =>  "products",
     "fields" => [
         { "name" => "name", "type" => "string" },
         { "name" => "available_months",  "type" => "string[]" },
         ........
       ]
    }
```
In the above schema, there is no way to filter such products that has no available months set
i.e available_months != [](empty array)

**Comment by kishorenc** [comment]
This would require more than just a simple `not equals` operator. It will require a length operation on the array field, and a check on that value. Something like `len(available_months): 0`. 

For now, a work around is to add a boolean field that captures the empty condition when the data is ingested, which you can then easily filter via an equals check.

**Comment by fxi** [comment]
And for logical connective such as `NOT <rule>` or `<attribute>!=<value>` or  `<attribute>!=[<value>, ...]` ? Doing this during ingestion is not a viable option, as it's not just `length=0`. 

Besides, in the documentation, it's not that clear to me, what logical or connective operators are allowed. I believe there is a mix of logical operators and array based logic, but the specs are not that explicit. Meilisearch is a bit more [clear on this](https://docs.meilisearch.com/reference/features/filtering.html#logical-connectives). It could be nice to have something similar.  At first, I though the filter string could be validated with something such as [lucene-query-parser](https://github.com/thoward/lucene-query-parser.js) but that's not the case. `typesense` seems to follow it's own specs. Maybe I'm wrong..

Thanks :)

**Comment by ronaldcurtis** [comment]
+1

My use case is that I have a `user_id_facet`, and I want to return results where the `user_id_facet` is not equal to the current user's id.

**Comment by kishorenc** [comment]
@ronaldcurtis 

The exclusion operator (-) is available on recent RC builds but only on string facets.

Usage:

```
filter_by=author:- JK Rowling
filter_by=author:- [JK Rowling, Enid Blyton]
```

Latest Docker RC build is: `typesense/typesense:0.21.0.rc15`

**Comment by kishorenc** [comment]
@ronaldcurtis I just tried this example on a `rc15` docker container and it works for me:

https://gist.github.com/kishorenc/052a0d25ddf62c1bd67c399c78abd13f

If you can produce a similar snippet with the issue, I will be happy to troubleshoot.

**Comment by ronaldcurtis** [comment]
@kishorenc Took me a while to figure it out, but it seems like it only works if `"num_memory_shards": 1,` is passed when creating the collection. Gist of working and not working snippets here: 

https://gist.github.com/ronaldcurtis/9c8995d0f86bf5b95958239792abc3ce

Thanks heaps for your help! Loving all these new features you're adding 🙏

**Comment by kishorenc** [comment]
@ronaldcurtis Good catch! I've figured out what's going wrong here. I will update this thread when I fix this issue and also tie up some loose ends as part of that.

**Comment by kishorenc** [comment]
@ronaldcurtis Can you please try against this new build:

`typesense/typesense:0.21.0.rc16`

**NOTE:** I've updated the operator for not equals to the conventional `!=` because later on, when we add support for numerical types, a negative sign (`-`) will be so confusing.

So, updated examples:

```
artist:!= Michael Jackson
artist:!= [Michael Jackson, Taylor Swift]
```

**Comment by ronaldcurtis** [comment]
@kishorenc Just tested it and its working great ^_^ Thanks heaps 👍 

> I've updated the operator for not equals to the conventional != because later on, when we add support for numerical types, a negative sign (-) will be so confusing.

That's a good point!
